templates/SynapseExternalBucket.yaml

```diff
@@ -123,6 +136,111 @@ Resources:
               - !If [AllowWrite, "s3:*Object*", "s3:GetObject*"]
               - "s3:*MultipartUpload*"
             Resource: !If [EnableEncryption, !Sub "${SynapseEncryptedExternalBucket.Arn}/*", !Sub "${SynapseExternalBucket.Arn}/*"]
+  ExternalBucketGroupPolicyUpdateRole:
+    Type: "AWS::IAM::Role"
+    Condition: CreateIPAddressRestrictionLambda
+    Properties:
+      AssumeRolePolicyDocument:
+        Version: '2012-10-17'
+        Statement:
+        - Effect: Allow
+          Principal:
+            Service: lambda.amazonaws.com
+          Action: sts:AssumeRole
+      Path: "/"
+      Policies:
+      - PolicyName: external-bucket-lambda-policy
+        PolicyDocument:
+          Version: '2012-10-17'
+          Statement:
+          - Effect: Allow
+            Action:
+            - logs:CreateLogGroup
+            - logs:CreateLogStream
+            - logs:PutLogEvents
+            Resource: "*"
+          - Action:
+            - s3:*
+            Effect: Allow
+            Resource: !If [EnableEncryption, !GetAtt SynapseEncryptedExternalBucket.Arn, !GetAtt SynapseExternalBucket.Arn]
+  ExternalBucketGroupPolicyUpdateLambda:
+    Type: "AWS::Lambda::Function"
+    Condition: CreateIPAddressRestrictionLambda
+    Properties:
+      Description: Dynamically updates the external bucket policy to restrict downloading only IP addresses listed by AWS.
+      Runtime: python3.6
+      Handler: "index.handler"
+      Role: !GetAtt ExternalBucketGroupPolicyUpdateRole.Arn
+      Timeout: 10
+      Code:
+        ZipFile: !Sub
+        - |
+          import json
+          import boto3
+          from botocore.vendored import requests
+
+          bucket_name = '${BucketName}'
+          region = '${Region}'
+          policy_statement_id="DenyGetObjectForNonMatchingIp"
+
+          def handler(event, context):
+            s3_client = boto3.client('s3')
+
+            #generate new policy statement based on data from AWS
+            ip_ranges = requests.get('https://ip-ranges.amazonaws.com/ip-ranges.json').json()['prefixes']
+            region_ip_addresses = [item['ip_prefix'] for item in ip_ranges if (item["service"] == "AMAZON" and item["region"] == region)]
+            
+            new_ip_policy_statement = {'Sid': policy_statement_id,
+                                       'Effect': 'Deny',
+                                       'Principal': '*',
+                                       'Action': 's3:GetObject',
+                                       'Resource': 'arn:aws:s3:::'+bucket_name+'/*',
+                                       'Condition': {'NotIpAddress': {'aws:SourceIp': region_ip_addresses}}}
```

Comment:
it looks like 'aws:SourceIP' need to be a list.  https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_examples_aws_deny-ip.html

Aren't you missing the brackets '[]' for list notation and commas between IP addresses?

Reply:
region_ip_addresses is already a list type. The brackets are automatically added by the json library.